### PR TITLE
Spring cloud gateways 2.1.x plugin

### PR DESCRIPTION
**Pull request by zhaoyuguang**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
https://github.com/apache/skywalking/issues/2126

### CODE REVIEW

**Comment by JaredTan95** [comment]
Looking forward test case :P

**Comment by JaredTan95** [comment]
In my opinion, this plugin could be split into two plugins: netty http plugin + spring cloud gateway plugin?

**Comment by zhaoyuguang** [comment]
> In my opinion, this plugin could be split into two plugins: reactor netty http plugin + spring cloud gateway plugin?

This also relate plugin which use `reactor-netty` such as webflux.

**Comment by wu-sheng** [comment]
@zhaoyuguang Is this plugin too complex and conflict with other plugins? If so, let's discuss how to do it first. I noticed this PR is still not finished.

**Comment by zhaoyuguang** [comment]
> @zhaoyuguang Is this plugin too complex and conflict with other plugins? If so, let's discuss how to do it first. I noticed this PR is still not finished.

In my opinion, there has been no impact so far, webflux enhances reactor-netty httpserver, gateway enhances reactor-netty httpclient, but I think it's best to separate reactor-netty, webflux, and gateway. Let users choose the supported version based on our testcase pass rate. Is it better?

**Comment by wu-sheng** [comment]
>  I think it's best to separate reactor-netty, webflux, and gateway. 

If you can do this, I think it is a good idea. If you could separate them, I would suggest you submit a PR to separate reactor-netty and webflux first, After it is landed, we work on gateway agian.

cc @ascrutae I think you should watch this.

**Comment by zhaoyuguang** [comment]
> > I think it's best to separate reactor-netty, webflux, and gateway.
> 
> If you can do this, I think it is a good idea. If you could separate them, I would suggest you submit a PR to separate reactor-netty and webflux first, After it is landed, we work on gateway agian.
> 
> cc @ascrutae I think you should watch this.

I also want to try it.

**Comment by wu-sheng** [comment]
You could try, I just bring Xin in the loop, he needs to review after you done.
 
Try not to make pr too big, that is helpful.

**Comment by zhaoyuguang** [comment]
> > I think it's best to separate reactor-netty, webflux, and gateway.
> 
> If you can do this, I think it is a good idea. If you could separate them, I would suggest you submit a PR to separate reactor-netty and webflux first, After it is landed, we work on gateway agian.
> 
> cc @ascrutae I think you should watch this.

I also want to try it.

> You could try, I just bring Xin in the loop, he needs to review after you done.
> 
> Try not to make pr too big, that is helpful.

ok

**Comment by wu-sheng** [comment]
We need to add witness class for this plugin, right?

**Comment by zhaoyuguang** [comment]
> We need to add witness class for this plugin, right?

Yes, I am still trying to be compatible with the low version

**Comment by wu-sheng** [comment]
OK. I has ping @ascrutae to deal with your other PRs.

**Comment by zhaoyuguang** [comment]
> OK. I has ping @ascrutae to deal with your other PRs.

OK

**Comment by wu-sheng** [comment]
Integration test fails, please verify, including the 2 2.1.x versions.

**Comment by zhaoyuguang** [comment]
I didn't find out why `actual data` has no trace data, it looks like it didn't make a http request, but it can't be reproduced in my local tesecase. @acurtain Can you help me? please.

**Comment by zhaoyuguang** [comment]
Some difficulties encountered:
1 testcase gateway 2.1.0 version, the result of trace data collection is not very stable, sometimes it will be missing, sometimes no problem
2 testcase gateway 2.1.1 version, the link data is synchronized, but I see that there is no change in the 2.1.0 version gateway code, and the results of my local test results are not the same.
image
image
@acurtain

**Comment by wu-sheng** [comment]
As the test tool can't support your PR. Could you provide a screenshot to show the result?

**Comment by wu-sheng** [comment]
Please change the title to `support Spring Gateway 2.1.x, and consider to create an issue to track `build plugin for SpringGateway 2.0.x`.

**Comment by zhaoyuguang** [comment]
> As the test tool can't support your PR. Could you provide a screenshot to show the result?

OK

File: docs/en/setup/service-agent/java-agent/Supported-list.md [hunk 1]
```diff
@@ -10,6 +10,7 @@
   * [Resin](http://www.caucho.com/resin-4.0/) 4 (Optional¹)
   * [Jetty Server](http://www.eclipse.org/jetty/) 9
   * [Spring Webflux](https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html) 5.x
+  * [Spring Cloud Gateway](https://spring.io/projects/spring-cloud-gateway) 2.1.x.RELEASE
```
Comment (wu-sheng) [inline docs/en/setup/service-agent/java-agent/Supported-list.md]: Please add optional tag.

**Comment by zhaoyuguang** [comment]
case1:common route
```yml
server:
  port: 8080
spring:
  cloud:
    gateway:
      routes:
      - id: provider_route
        uri: http://${projectB.host}
        predicates:
        - Path=/provider/b/*
```
image
image

case2:eureka registration center
```yml
server:
  port: 8085
spring:
  cloud:
    gateway:
      discovery:
        locator:
          enabled: true
          lower-case-service-id: true
  application:
    name: spring-cloud-gateway
eureka:
  client:
    service-url: 
      defaultZone: http://localhost:8081/eureka/
```
image

File: docs/en/setup/service-agent/java-agent/Supported-list.md (hunk 1 above)
Comment (zhaoyuguang) [reply docs/en/setup/service-agent/java-agent/Supported-list.md]: ok,  move `Spring Cloud Gateway` at `HTTP Server` to `HTTP Gateway` is better?

**Comment by wu-sheng** [comment]
Is the topology right? Why Handler span is not the parent of exit span?

**Comment by wu-sheng** [comment]
This trace looks unclear to me. 

For a gateway, the ideal spans at least are
1. Entry span of gateway.
2. Exit Span to callong upstream system
3. Entry span is the parent of exit span.

Can you explain a little more? And show me the topology?

**Comment by wu-sheng** [comment]
Also, I noticed the handler is a local span, this local span is created by gateway plugin, right? If you put route info in it, then, make sure, exit span is a children of it.

**Comment by zhaoyuguang** [comment]
image

**Comment by zhaoyuguang** [comment]
After some mins, the topo page will be 
image

**Comment by zhaoyuguang** [comment]
> Is the topology right? Why Handler span is not the parent of exit span?

The topo page like this.
image

The handle span is `spring-cloud-gateway` handle the entry API route, Which base on config such as yml

**Comment by zhaoyuguang** [comment]
> This trace looks unclear to me.
> 
> For a gateway, the ideal spans at least are
> 
> 1. Entry span of gateway.
> 2. Exit Span to callong upstream system
> 3. Entry span is the parent of exit span.
> 
> Can you explain a little more? And show me the topology?


image

**Comment by zhaoyuguang** [comment]
> Also, I noticed the handler is a local span, this local span is created by gateway plugin, right? If you put route info in it, then, make sure, exit span is a children of it.

Ok, I think about it.

**Comment by zhaoyuguang** [comment]
I remove create local span and add route tag at active span
image
image

**Comment by wu-sheng** [comment]
Thanks for providing https://github.com/apache/skywalking/pull/2718 to fix the related issue.

**Comment by zhaoyuguang** [comment]
> Thanks for providing #2718 to fix the related issue.

Also, I can study more.

**Comment by wu-sheng** [comment]
The endpoint bug fixed. Please update Gateway.

**Pull request merged**